src/libertem/udf/base.py

```diff
@@ -606,20 +656,43 @@ def get_preferred_input_dtype(self):
         return np.float32
 
     def get_tiling_preferences(self):
+        '''
+        .. versionadded:: 0.6.0.dev0
+        '''
         return {
             "depth": UDF.TILE_DEPTH_DEFAULT,
             "total_size": UDF.TILE_SIZE_MAX,
         }
 
+    def get_backends(self):
+        # TODO see interaction with C++ CUDA modules requires a different type than CuPy
+        '''
+        Signal which computation back-ends the UDF can use.
+
+        :code:`numpy` is the default CPU-based computation.
+
+        :code:`cupy` is CUDA-based computation through CuPy.
```

Comment:
I think, the CUDA worker should named `cuda`, because you can use every CUDA-based library.

Reply:
In principle, yes. That's why the more low-level interfaces call it "cuda" and not "cupy", for example the resource tags in the cluster. At the level of UDFs, some things are cupy-specific:

* self.xp is set to the cupy module
* tiles/frames/partitions are passed as cupy arrays
* Result buffers are allocated as cupy arrays if `where="device"`
* In the future, datasets will likely read cupy arrays directly

In the future, CuPy may get support for other devices, such as AMD, or we might want to allow using other interfaces than cupy, for example a different device array library. For that reason the UDF specifies the interface it uses, and not the back-end on which the library runs.

Reply:
Okay, but in the case of using the alpaka-pybind11-bindings, just the CUDA environment is necessary and not the `cupy` library. So there is already a use-case, where `cupy` is not required.

Reply:
Yes, good point. Is using CuPy arrays for handling data exchange with C++ CUDA code suboptimal or is it a significant burden to install CuPy? In that case adding support for a third back-end would make sense.

Reply:
CuPy should not affect the data exchange with CUDA C++. The exchange is done by numpy arrays, which are also the base for CuPy. 

But, I see another problem. CuPy needs to install a package depending on the version of the CUDA toolkit (in my case 10.2). I think, this behaviour make it hard to automate the installation and the package is really huge. In my case about 340 MB. So, I'm not sure, if CuPy should required for CUDA. But in general, it does not cause any problem with the bindings.

Reply:
Ok, I've rewritten the CUDA support in such a way that it should work completely without CuPy. By requesting backend "cuda", a UDF is scheduled on CUDA workers, but otherwise behaves as a regular NumPy UDF, i.e. no CuPy result buffers, input data etc. That is then left to the user. :-)

Reply:
Sounds good. I will test it with the alpaka-bindings in the next days.